program_intent_engagement/apps/api/v1/views.py

```diff
@@ -1 +1,89 @@
-# Create your views here.
+"""
+V1 API Views
+"""
+import logging
+
+from django.core.checks import Error
+from django.core.exceptions import ValidationError
+from edx_rest_framework_extensions.auth.jwt.authentication import JwtAuthentication
+from rest_framework import status
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from program_intent_engagement.apps.api.serializers import ProgramIntentSerializer
+from program_intent_engagement.apps.core.models import ProgramIntent
+
+log = logging.getLogger(__name__)
+
+
+class ProgramIntentsView(APIView):
+    """
+    View to insert program intents.
+
+    Given a valid program intent, this view will insert the program intent. If the given program intent already
+    exists, no action is taken (no duplicate row).
+
+    Path: /api/[version]/program_intents
+
+    Accepts: [POST]
+
+    HTTP POST: Inserts a Program Intent
+
+    Expected POST data: {
+        "program_uuid": "949d5e37-e16c-4e95-9104-8c2807caa8ce",
+        "reason": "ALL_COURSES_COMPLETE",
+        "certainty": "CERTAIN_YES",
+        "effective_timestamp": "2022-07-01 00:00:00"
+    }
+
+    POST data parameters:
+     * program_uuid: The uuid of the program intent.
+     * reason: The reason for how intent was measured.
+     * certainty: The certainty of the intent for the user.
+     * effective_timestamp: When the event occurred that created the intent.
+    Returns:
+     * 200: The program intent was inserted. If duplicate, no action was taken.
+     * 400: The program intent request was invalid.
+    """
+
+    authentication_classes = (JwtAuthentication,)
+    permission_classes = (IsAuthenticated,)
+
+    @classmethod
+    def add_program_intent_lms_user_id(cls, request):
+        """
+        Add lms_user_id to program intent serializer from request user
+        """
+        # get lms user id from the request user and add to program intent request
+        request.data["lms_user_id"] = request.user.lms_user_id
+
+        return ProgramIntentSerializer(data=request.data)
+
+    def post(self, request):
+        """
+        Create and insert a program intent. Duplicates not inserted.
+        """
+        serializer = self.add_program_intent_lms_user_id(request)
+
+        if serializer.is_valid():
+            # check if program intent exists, create if it does not exist
+            # program_intent = ProgramIntent.objects.filter(**request.data)
+            #
+            # # todo: update logic depending on discussion
+            # if program_intent.exists():
+            #     response_status = status.HTTP_200_OK
+            #     data = {"detail": "This program intent already exists"}
+            # else:
+            serializer.save()
+            response_status = status.HTTP_200_OK
+            data = {}
+        else:
+            if serializer.errors["non_field_errors"][0].code == "unique":
```

Comment:
After some investigating and testing, it looks like the serializer validation is actually what is surfacing the 400 Validation Error based on failing the unique constraint. So, as I'd discussed with @alangsto, there are two potential approaches to suppressing the unique constraint serializer ValidationError:
1) The current approach of parsing the serialization error. This needs to be tweaked slightly to handle the cases when the non_field_errors don't exist (i.e. for missing required fields), so I do need to change that. The con is that this is a little messy.
2) Another possible approach is disabling the uniqueness constraint in the serializer as discussed [here](https://stackoverflow.com/questions/45999131/django-rest-framework-model-serializer-with-out-unique-together-validation). As Alie mentioned, the main con here is that this could lead to unexpected behavior especially if the serializer is used everywhere.

My instinct is that 1 is the safer approach but wanted to discuss. Also wanted to see if there could be another approach altogether. @ashultz0

Reply:
The fact that the serializer ACCESSES THE DATABASE is a betrayal of basically everything a serializer is supposed to do and I'm going to dedicate a gin and tonic to it.

Messing with the validators is delicate and a future bug generator, I agree it's not ideal.
Parsing the serialization error is also a huge bug generator.

The nice behavior of not throwing an error when someone makes duplicate calls because of confusion is not worth planting either of those bug seeds. Just go back to throwing the 400 and we'll document that behavior.

Reply:
Makes sense, @ashultz0  (your suggestions, not the serializer accessing the database part haha, that's just silly!). 

Just made that change. I updated the docs, but is there somewhere specific you were suggesting to document? I'm guessing the swagger-generated api-docs?

Reply:
we haven't written the docs yet, it can go into Allie's ticket